spec/index.bs

```diff
@@ -1270,10 +1298,12 @@ The first few steps of the [=prepare the script element=] algorithm are modified
     run when the parser tries to run it, but it is later executed after a script dynamically
     updates it, it will execute in an async fashion even if the <code id=script-processing-model:attr-script-async-5><a href=https://html.spec.whatwg.org/#attr-script-async>async</a></code> attribute isn't set.</p>
   <li><ins><p>Execute the [$Prepare the script text$] algorithm on <var>el</var>. If that algorithm threw an error, then return.</p></ins></li>
-  <li><p>Let <var ignore="">source text</var> be <var>el</var>'s <del><a id=script-processing-model:child-text-content href=https://dom.spec.whatwg.org/#concept-child-text-content data-x-internal=child-text-content>child text content</a>.</del> <ins>[=script text=] value.</ins>
+  <li><p>Let <var ignore="">source text</var> be <var>el</var>'s <del><a id=script-processing-model:child-text-content href=https://dom.spec.whatwg.org/#concept-child-text-content data-x-internal=child-text-content>child text content</a>.</del> <ins>[=HTMLScriptElement/script text=] value.</ins>
   <li>...
   </ol>
 
+Issue: There's no proper definition for the processing of SVG script elements. However, you should apply a similar change to the processing of {{SVGScriptElement}}s.
```

Comment:
@annevk can you think of a better way to word this, or a way to actually spec this?

I could try patching https://www.w3.org/TR/SVGMobile12/script.html#ScriptContentProcessing but it's a bit lose on the details and is fairly old, so idk that we'd ever upstream the patching.

Reply:
The boolean bit is currently only used as part of the "have the children been changed by an API during parsing" code.

Reply:
The problem with a Boolean bit is how to allow the specific IDL and the "actually parsed" stuff but nothing else. I'm currently unable to think how you would do that.

Reply:
Everything else that mutates would end up flipping the bit? It might help if we went through all the entry points we are concerned about or why this would be hard for some of them.

Reply:
E.g., is the tradeoff that some mutation algorithm would need to have knowledge about script elements? And what do implementations do here? I find it hard to believe they would accept doubling the cost of each script element.

Reply:
So we want to allow when the parser gets to `</script>` and the contents haven't already been touched by JS.

We also want to allow HTMLScriptElement.text HTMLScriptElement.textContent and HTMLScriptElement.innerText to allow setting the contents (once the script is parsed).

So examples of what we want to fail are stuff like appendChild(textNode), or Node.textContent setter being accessed weirdly with the script element.

So is your idea to update basically every callsite with an "if script flip trusted bit to false"? I think my worry is that we'd want that change to be deeper than callsites and then it's a matter of how to say this mutation is okay and this one isn't.

Chromium right now does the store a second copy of contents and only touches the 3 IDL setters, the parser end step and children changed steps.

Reply:
I'm a bit confused by the question so I think I'm misunderstanding something here.

Yes we want to track if it's an API or parser triggering children changed steps. So currently this is used for the weird API change mid-parse case.

If we used a boolean for trust rather than duplicating the internals we could mark is as trusted in parser and that would be fine.

But after parse is finished this would be triggered by all API changes most of which we don't want to trust. But it would also be triggered by HTMLScriptElement.text for example. How would we allow these sanctioned APIs?

If we do HTMLScriptElement.text setter sets the trusted bool then that's fine but it would then trigger children changed steps as an API and we'd have to handle that (presumably the idea would be to mark it as untrusted?)

Reply:
In the cases where we trust the API can't we just set the bit to trusted after the children changed steps have run? I suppose with mutation events that might not quite work. But they're close to disappearing. Hmm.

Reply:
Yeah I did consider something like that where we have a boolean to say this is an approved API change and then in children changed we use that to set trusted (so we'd have maybe 3 booleans?), but I felt slightly uneasy about it.

It's possible though that in practice any shenanigans would result in an untrusted script.

Reply:
I'm still at 1 boolean. Presumably if a script set is trustworthy you can just set the existing bit to trustworthy in the end.

Reply:
Ah yeah actually we don't need a parse specific one anymore and provided it's all synchronous you're right you could set the "trusted" bit to true after the children changed steps (which would set it to false for all API calls).

If we make children changed always untrust API changes I think that solves any weirdness from mutation events too. Because we'd fail safe and the script would be untrusted. Because we only the 3 sanctioned APIs to behave properly wrt to mutation events.

Thinking on it some more I think that could work?

cc @otherdaniel what do you think about that (Assuming we can get children changed steps spec to correctly tell us if it's API or parser)?

Relatively small changes needed (assuming my mental model is right) such as that below and then updating the children changed steps and the prepare script steps:

<img width="1485" alt="image" src="https://github.com/w3c/trusted-types/assets/32498324/d21cba36-1571-4b19-9eea-ea14a7dc80c9">

Reply:
Thinking on this some more it would also fix https://github.com/w3c/trusted-types/issues/517 (different handling of new lines leading to TT errors)

Reply:
Okay so I tried implementing this and hit a snag.

https://html.spec.whatwg.org/#prepare-the-script-element:~:text=When%20a%20script%20element%20el%20that%20is%20not%20parser%2Dinserted%20experiences

This bit of the HTML spec is ran inside of children changed steps (at least in WebKit and I believe Chromium from a quick look), so prepare the script runs *before* the isTrusted flag is set back to true at the end of setInnerText.

Reply:
I've opened a new draft PR which I'll use to do this new model (and will also apply to svg as best it can) https://github.com/w3c/trusted-types/pull/533